### PR TITLE
plan: handle DNF expressions in `Selectivity`

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/8921

### What is changed and how it works?

Consider DNF expressions when extracting ranges for estimation in `Selectivity`.

### Check List 

Tests 

 - Integration test

Code changes

N/A

Side effects

N/A

Related changes

N/A

### CODE REVIEW

File: util/ranger/detacher.go [hunk 1]
```diff
@@ -298,10 +298,9 @@ func detachDNFCondAndBuildRangeForIndex(sctx sessionctx.Context, condition *expr
 }
 
 // DetachCondAndBuildRangeForIndex will detach the index filters from table filters.
-// If the top layer is DNF, we return a int slice which is eqAndInCount of every DNF item.
-// Otherwise just one number is returned.
+// If the top layer is CNF, we return a int of eqAndInCount, otherwise we return 0.
 func DetachCondAndBuildRangeForIndex(sctx sessionctx.Context, conditions []expression.Expression, cols []*expression.Column,
-	lengths []int) ([]*Range, []expression.Expression, []expression.Expression, int, error) {
+	lengths []int) (ranges []*Range, accesses, filters []expression.Expression, eqCount int, isDNF bool, err error) {
```
Comment (shenli) [inline util/ranger/detacher.go]: This function returns 6 things. Could we merge the first five things into a structure?

Comment (eurekaka) [reply util/ranger/detacher.go]: Done.

File: statistics/selectivity.go [hunk 2]
```diff
@@ -251,18 +251,25 @@ func (coll *HistColl) Selectivity(ctx sessionctx.Context, exprs []expression.Exp
...
+		res, err = ranger.DetachCondAndBuildRangeForIndex(ctx, exprs, cols, lengths)
+		ranges, accessConds, remainedConds, isDNF = res.Ranges, res.AccessConds, res.RemainedConds, res.IsDNFCond
 	default:
 		panic("should never be here")
 	}
 	if err != nil {
-		return 0, nil, errors.Trace(err)
+		return 0, nil, err
+	}
+	if isDNF && len(remainedConds) == 0 {
```
Comment (lamxTyler) [inline statistics/selectivity.go]: So if it has remained conds, the selectivity is still `row count` * 0.8?

File: util/ranger/detacher.go [hunk 3]
```diff
@@ -297,11 +309,20 @@ func detachDNFCondAndBuildRangeForIndex(sctx sessionctx.Context, condition *expr
...
+	AccessConds   []expression.Expression
+	RemainedConds []expression.Expression
+	EqCondCount   int
+	IsDNFCond     bool
+}
+
 // DetachCondAndBuildRangeForIndex will detach the index filters from table filters.
-// If the top layer is DNF, we return a int slice which is eqAndInCount of every DNF item.
-// Otherwise just one number is returned.
+// If the top layer is CNF, we return a int of eqAndInCount, otherwise we return 0.
```
Comment (lamxTyler) [inline util/ranger/detacher.go]: Update comment?

File: statistics/selectivity.go (hunk 2 above)
Comment (eurekaka) [reply statistics/selectivity.go]: Updated, PTAL

File: statistics/selectivity.go [hunk 4]
```diff
@@ -240,6 +243,9 @@ func (coll *HistColl) Selectivity(ctx sessionctx.Context, exprs []expression.Exp
 	for _, set := range usedSets {
 		mask &^= set.mask
 		ret *= set.Selectivity
+		if set.partCover {
+			ret *= selectionFactor
```
Comment (zz-jason) [inline statistics/selectivity.go]: could you add a comment to explain why we multiple `selectionFactor`?

Comment (eurekaka) [reply statistics/selectivity.go]: Done.

**Comment by zz-jason** [comment]
@eurekaka We'd better to cherry pick this commit to release 2.1.

**Pull request merged**